pallets/xvm/src/lib.rs

```diff
@@ -120,25 +122,53 @@ where
         value: Balance,
         skip_execution: bool,
     ) -> CallResult {
+        let overheads = match vm_id {
+            VmId::Evm => WeightInfoOf::<T>::evm_call_overheads(),
+            VmId::Wasm => WeightInfoOf::<T>::wasm_call_overheads(),
+        };
+
         ensure!(
             context.source_vm_id != vm_id,
             CallErrorWithWeight {
-                error: CallError::SameVmCallNotAllowed,
-                used_weight: match vm_id {
-                    VmId::Evm => WeightInfoOf::<T>::evm_call_overheads(),
-                    VmId::Wasm => WeightInfoOf::<T>::wasm_call_overheads(),
-                },
+                error: CallError::SameVmCallDenied,
+                used_weight: overheads,
             }
         );
 
-        match vm_id {
-            VmId::Evm => {
-                Pallet::<T>::evm_call(context, source, target, input, value, skip_execution)
-            }
-            VmId::Wasm => {
-                Pallet::<T>::wasm_call(context, source, target, input, value, skip_execution)
-            }
+        // Set `IN_XVM` to true & check reentrance.
+        if IN_XVM.with(|in_xvm| in_xvm.replace(true)) {
+            return Err(CallErrorWithWeight {
+                error: CallError::ReentranceDenied,
+                used_weight: overheads,
+            });
```

Comment:
Do we need to clear `IN_XVM` here also? The XVM call should terminate here but asking just in case

Reply:
The XVM call terminates here. After I finished the next ticket *propagating XVM errors*, will update the integration test to assert the `ReentranceDenied` error. It will be obvious to verify this by then.

Reply:
Got it!